renv::restore()

library(shiny)
library(ggplot2)



shinyUI(navbarPage("DECoN",


	tabPanel("Introduction",
#		progressInit()
                h2('Welcome!'),
                p('This web app provides an interface to the results of running DECoN on your data. You can check the data which has been used, evaluate the coverage of individual exons and individual samples and visualise the CNV calls.'),
                p('The data shown is from the input file provided, generated by DECoN version 1.0.1. If it\'s not what you\'re expecting, then first check that this is the correct file and that it has been generated according to the documentation.'),
                p('This program requires that you have a modern internet browser (Firefox, Chrome or IE10) and R installed.')
        ),
	

        tabPanel("Data",
            sidebarLayout(
                sidebarPanel(
                    h3("Show/hide files and details"),
    	            actionButton("readBed","BED file"),
                    p(),
	            actionButton("readBams","BAM files and sample names"),
                    p(),
	            actionButton("readfasta","FASTA file")
                ),

                mainPanel(
                    h3("BED file"),
	            dataTableOutput("bed.file"),
                    h3("Samples"),
	            dataTableOutput("bamlist"),
	            h3("Reference file "), 
	            textOutput("fasta")
	        )
            )
        ),



        tabPanel("Coverage Evaluation",
	
            downloadButton("downloadCov", "Download Coverages"),
            h3('Failed Samples'),

            sidebarLayout(
                sidebarPanel(
                    numericInput("SampleCovLimit",min=0,value=50,label="Median coverage threshold for failed samples"),
                    numericInput("SampleCorLimit",min=0,max=1,value=.98,label="Correlation threshold for failed samples",step=.01)
                ),
                mainPanel(
                    dataTableOutput("FailedSamples")
                )
            ),
            
            h3("Failed Exons"),
            
            sidebarLayout(
                sidebarPanel(
                    numericInput("ExonCovLimit",min=0,value=50,label="Median coverage threshold for failed exons")
                ),
                mainPanel(
                    dataTableOutput("FailedExons")
                )
            ),

            h3("Plot of Coverage"),
        
            sidebarLayout(
                sidebarPanel(
                    h5('Sample to plot in blue:'),
                    uiOutput("PlotSampleHighlight"),
                    h5('Additional Samples to plot:'),
                    radioButtons("ChooseFrom",choices=list("Reference Samples"=1,"All Samples"=2),label="Choose From: "),
                    uiOutput("PlotSamplesInput"),
                    h5('Genes to plot:'),
                    uiOutput("PlotGenesInput"),
                    h5('Plotting options:'),
                    radioButtons("plotType",choices=list("Box plot"=1,"Line plot"=2),label="Type of plot"),
                    radioButtons("plotScale",choices=list("None"=1,"Log"=2),label="Data transformation")
                ),
                mainPanel(
                    plotOutput("CovPlot",height=700)
                )
            )
        ),





        tabPanel("CNV Calls",
	    h2("CNV Calls"),
            downloadButton("downloadCNVs", "Download CNVs"),
            p(),
            dataTableOutput("CNVcalls"),


            fluidRow(
            column(3,
                h4("Select a variant"),
                uiOutput("selVar")
            ),
            column(3,
                h4("Change Plotted Exons"),
                uiOutput("minEx"),
                uiOutput("maxEx")
            ),
            column(3,
                h4("Change additional samples to be plotted"),
                radioButtons("chSamp",choices=list("Reference"=1,"All"=2,"None"=3),label="Select set of samples to be plotted in addition to the sample containing the variant")
            ),
            column(3,
                h4("Scale"),
                radioButtons("chScale",choices=list("Raw data"=1,"Log transformed"=2),label="Plot raw data or data on log scale")
                #,"By exon (Not available when plotting no additional samples)"=3,"By sample"=4,"By exon and sample (Not available when plotting no additional samples)"=5),label="Transform data or normalise by subtracting median.")
            )
        ),

        plotOutput('plot',height=200),
        plotOutput('genes',height=75),#,width="87.5%"),
        plotOutput('CIplot',height=150)#,width="87.5%")

    )#,






   # tabPanel("Details")
    )
)
